File: Assets/Tests/TestExceptions.cs
using System.Threading.Tasks;

namespace Nahoum.UnityJSInterop.Tests
{
    public class TestExceptions
    {
        [ExposeWeb]
        public static void ThrowSimpleException()
        {
            throw new System.Exception(SampleValues.TestExceptionValue);
        }

        [ExposeWeb]
        public static void TestUnraisedException()
        {
            string obj = null;
            obj.ToString();
        }

        [ExposeWeb]
        public static async Task ThrowSimpleExceptionAsync()
        {
            await Task.Yield();
            throw new System.Exception(SampleValues.TestExceptionValue);
        }

        [ExposeWeb]
        public static async Task TestUnraisedExceptionAsync()
        {
            await Task.Yield();
            string obj = null;
            obj.ToString();
        }
    }
}

File: sample-unity-js-interop-ts/src/tests/test-exceptions.ts
import { UnityInstance } from "../UnityJSInterop";
import { assertThrows, assertThrowsAsync } from "../test-utilities/test-asserts";

export async function RunTestExceptions(unityInstance: UnityInstance){
    const instance = unityInstance.Module.static.Nahoum.UnityJSInterop.Tests.TestExceptions;
    assertThrows(() => instance.ThrowSimpleException(), "Simple exception");
    assertThrows(() => instance.TestUnraisedException(), "Complex exception");
    await assertThrowsAsync(() => instance.ThrowSimpleExceptionAsync(), "Simple exception async");
    
    // Not supported yet
    //await assertThrowsAsync(() => instance.TestUnraisedExceptionAsync(), "Complex exception async");

}